applications/nrf_desktop/src/hw_interface/selector_hw.c

```diff
@@ -0,0 +1,325 @@
+		err = gpio_pin_disable_callback(gpio_dev[sel_pins[i].port],
+						sel_pins[i].pin);
+
+		if (err) {
+			LOG_ERR("Cannot disable callback (err %d)", err);
+			break;
+		}
+	}
+
+	return err;
+}
+
+static void selector_isr(struct device *dev, struct gpio_callback *cb,
+			 u32_t pins_mask)
+{
+	u8_t port = dev - gpio_dev[0];
+	struct selector *sel;
+
+	sel = CONTAINER_OF((u8_t *)cb - port * sizeof(struct gpio_callback),
+			   struct selector,
+			   gpio_cb);
+	disable_interrupts_nolock(sel);
+
+	k_delayed_work_submit(&sel->work, DEBOUNCE_DELAY);
+}
+
+static int read_state_and_enable_interrupts(struct selector *selector)
+{
+	int err = read_state(selector);
+
+	if (err) {
+		return err;
+	}
+
+	err = enable_interrupts_lock(selector);
+
+	return err;
+}
+
+void selector_work_fn(struct k_work *w)
+{
+	struct selector *selector = CONTAINER_OF(w, struct selector, work);
+
+	int err = read_state_and_enable_interrupts(selector);
+
+	if (err) {
+		module_set_state(MODULE_STATE_ERROR);
+	}
+}
+
+static int configure(struct selector *selector)
+{
+	const struct gpio_pin *sel_pins = selector->config->pins;
+	u32_t bitmask = 0;
+
+	for (size_t i = 0; i < ARRAY_SIZE(gpio_dev); i++) {
+		for (size_t j = 0; j < selector->config->pins_size; j++) {
+			if (sel_pins[j].port == i) {
+				bitmask |= BIT(sel_pins[j].pin);
+			}
+		}
+		gpio_init_callback(&selector->gpio_cb[i], selector_isr,
+				   bitmask);
+		int err = gpio_add_callback(gpio_dev[i], &selector->gpio_cb[i]);
+
+		if (err) {
+			LOG_ERR("Cannot add callback (err %d)", err);
+			return err;
+		}
+
+		bitmask = 0;
+	}
+
+	for (size_t i = 0; i < selector->config->pins_size; i++) {
+		int err = gpio_pin_configure(gpio_dev[sel_pins[i].port],
+					     sel_pins[i].pin,
+					     GPIO_PUD_PULL_DOWN | GPIO_DIR_IN |
+					     GPIO_INT | GPIO_INT_LEVEL |
+					     GPIO_INT_ACTIVE_HIGH);
+
+		if (err) {
+			LOG_ERR("Cannot configure pin (err %d)", err);
+			return err;
+		}
+	}
+
+	k_delayed_work_init(&selector->work, selector_work_fn);
+
+	return 0;
+}
+
+static int init(void)
+{
+	BUILD_ASSERT_MSG(ARRAY_SIZE(selector_config) <= UCHAR_MAX,
+			 "Maximum number of selectors is 255");
+	for (size_t i = 0; i < ARRAY_SIZE(port_map); i++) {
+		gpio_dev[i] = device_get_binding(port_map[i]);
+
+		if (!gpio_dev[i]) {
+			LOG_ERR("Cannot get GPIO%u dev binding", i);
+			return -ENXIO;
+		}
+	}
+
+	int err;
+
+	for (size_t i = 0; i < ARRAY_SIZE(selectors); i++) {
+		selectors[i].config = &selector_config[i];
+		err = configure(&selectors[i]);
+		if (err) {
+			return err;
+		}
+		err = read_state_and_enable_interrupts(&selectors[i]);
+		if (err) {
+			return err;
+		}
+	}
+	state = STATE_ACTIVE;
+
+	return 0;
+}
+
+static void sleep(void)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(selectors); i++) {
+		int err = disable_interrupts_nolock(&selectors[i]);
+
+		if (err) {
+			module_set_state(MODULE_STATE_ERROR);
+			return;
+		}
+
+		k_delayed_work_cancel(&selectors[i].work);
```

Comment:
Is that ok, if a work is pending when we go into error state?

Reply:
After propagating information about error state by power_manager module should go to sleep (as reaction for `power_down_event`)